### PR TITLE
Feature/698 Machine Learning and Azure ML Profiling 

### PR DESCRIPTION
**Pull request by vJenny**
# #698 

## Description

---

This PR contains a document for the ML Fundamentals section describing some common scenarios in MLOps/Data Science projects, along with suggestions on how to profile them. This document also includes recipes for the following scenarios: generic profiling, deep learning solution profiling, AzureML pipelines profiling. 

## PR Checklist

---

- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] My code follows the code style of this project.
- [x] I ran the lint checks which produced no new errors nor warnings for my changes.
- [x] I have checked to ensure there aren't other open Pull Requests for the same update/change.

## Does this introduce a breaking change?

---

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by vJenny** [comment]
`Markdownlint / markdown-link-check (pull_request)` is failing on a file that I've never touched. Any suggestions here?

File: docs/machine-learning/ml-profiling.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
+# Profiling Data Science and MLOps Code
```
Comment (omri374) [inline docs/machine-learning/ml-profiling.md:1]: Suggestion: Rephrase to "Profiling ML and MLOps Code". The term "Data Science Code" might be ambiguous to some readers.

File: docs/machine-learning/ml-profiling.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+- [Profiling Data Science and MLOps Code](#profiling-data-science-and-mlops-code)
+  - [Generic Python profiling](#generic-python-profiling)
+  - [PyTorch model training profiling](#pytorch-model-training-profiling)
+  - [Azure Machine Learning pipeline profiling](#azure-machine-learning-pipeline-profiling)
+
+## Generic Python profiling
+
+Usually an MLOps/Data Science solution contains plain Python code serving different purposes (e.g., data processing) along
+with specialized model training code. Whereas many Machine Learning frameworks provide their own profiler,
```
Comment (omri374) [inline docs/machine-learning/ml-profiling.md:18]:
```suggestion
with specialized model training code. Although many Machine Learning frameworks provide their own profiler,
```

not an English expert :) this sounded a lit more natural to me.

File: docs/machine-learning/ml-profiling.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+sometimes it is also very useful to profile the whole solution.
+
+There are two types of profilers: deterministic (all events are tracked) and statistical (sampling with regular intervals). In this
+document we'll cover an example of a deterministic profiler.
+
+There are many options of generic deterministic Python code profiling. The default option for profiling used to be a built-in
```
Comment (omri374) [inline docs/machine-learning/ml-profiling.md:24]: You mention that cProfile *used to be* the default option. From [this doc](https://docs.python.org/3/library/profile.html) it seems that both `cProfile` and `profile` are considered default. 

Maybe consider saying that cProfile is a suggested / popular / common profiling tool.

File: docs/machine-learning/ml-profiling.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+[cProfile](https://docs.python.org/3/library/profile.html) profiler. Using *cProfile* one can easily profile
+either a Python script or just a chunk of code. This profiling tool produces a file that can be either
+visualized using open source tools or analyzed using `stats.Stats` class. The latter option requires setting up filtering
+and sorting parameters for better analysis experience.
+
+Below you can find an example of using *cProfile* to profile a chunk of code.
+
+```python
+import cProfile
```
Comment (omri374) [inline docs/machine-learning/ml-profiling.md:33]: There's also the option of running it from the outside (`python -m cProfile [-o output_file] [-s sort_order] (-m module | myscript.py)`). Is this something worth considering here? Just a thought, not sure as the code is likely to be ran remotely and not by the user.

File: docs/machine-learning/ml-profiling.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+profiler.disable()
+
+# Write profiler results to an html file
+profiler.dump_stats("profiler_results.prof")
+```
+
+> Note: one epoch of model training is usually enough for profiling. There's no need to run more epochs and produce
+additional cost.
+
+More information on *cProfiler*:
```
Comment (omri374) [inline docs/machine-learning/ml-profiling.md:51]:
```suggestion
More information on *cProfile*:
```

File: docs/machine-learning/ml-profiling.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    # Automatically saves profiling results to disk
+    on_trace_ready=torch.profiler.tensorboard_trace_handler,
+    with_stack=True
+) as profiler:
+    for step, data in enumerate(trainloader, 0):
+        # -- TRAINING STEP CODE GOES HERE ---
+        profiler.step()
+```
+
+The tensorboard_trace_handler for analysis in TensorBoard. Those can be visualized by installing TensorBoard
```
Comment (omri374) [inline docs/machine-learning/ml-profiling.md:77]:
```suggestion
The `tensorboard_trace_handler` can be used to generate result files for TensorBoar. Those can be visualized by installing the TensorBoard
```

File: docs/machine-learning/ml-profiling.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+- [PyTorch Profiler Recipe](https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html)
+- [Introducing PyTorch Profiler - the new and improved performance tool](https://pytorch.org/blog/introducing-pytorch-profiler-the-new-and-improved-performance-tool/)
+
+## Azure Machine Learning pipeline profiling
+
+In our projects we often use [Azure Machine Learning](https://azure.microsoft.com/en-us/services/machine-learning/)
+pipelines to train Machine Learning models. Most of the profilers can also be used in conjunction with Azure Machine Learning.
+In a nutshell, you need to choose a profiler, that meets the following criteria:
+
+- The profiler can be enabled and disabled inside of the code
```
Comment (omri374) [inline docs/machine-learning/ml-profiling.md:99]: This isn't very clear. Do you mean that one can control whether the profiler is on or off using a parameter, or that certain parts of the code would be profiled and others won't? Consider rephrasing.

File: docs/machine-learning/ml-profiling.md [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+- The profiler produces a file as an output
+
+The general recipe of using profilers with Azure Machine Learning is the following:
```
Comment (omri374) [inline docs/machine-learning/ml-profiling.md:102]:
```suggestion
The general, a recipe for using profilers with Azure Machine Learning is the following:
```

**Review by omri374 — COMMENTED** [review]
Great contribution! Added a few comments but overall looks great. Thanks Yana

**Comment by omri374** [comment]
@vJenny regarding the failing markdown test, it's a dead link. A quick fix could be to remove the link. If you feel this would require more changes to the content, I would suggest to reach out to the original author of this page.
```
FILE: ./docs/automated-testing/shadow-testing/README.md
[✖] https://www.envoyproxy.io
```

File: docs/machine-learning/ml-profiling.md (hunk 1 above)
Comment (vJenny) [reply docs/machine-learning/ml-profiling.md]: Updated, thanks!

File: docs/machine-learning/ml-profiling.md (hunk 3 above)
Comment (vJenny) [reply docs/machine-learning/ml-profiling.md]: Rephrased to _"One of the default options for profiling..."_, thanks!

File: docs/machine-learning/ml-profiling.md (hunk 4 above)
Comment (vJenny) [reply docs/machine-learning/ml-profiling.md:32]: Added, thanks a lot for the suggestion!

File: docs/machine-learning/ml-profiling.md (hunk 5 above)
Comment (vJenny) [reply docs/machine-learning/ml-profiling.md]: Reworded using @PreteshPatel 's  suggestion, thanks!

File: docs/machine-learning/ml-profiling.md [hunk 9]
```diff
@@ -0,0 +1,120 @@
+# Profiling Data Science and MLOps Code
+
+Data Science projects, especially the ones that involve Deep Learning techniques, usually are very resource intensive. One model training iteration might be multiple hours long. Although large data volumes processing genuinely takes time, minor bugs and nonoptimal implementation of some functional pieces might cause extra resources consumption.
+
+Profiling can be used to identify performance bottlenecks and see which functions are the costliest in the application code. Based on the outputs of the profiler, one can focus on largest and easiest-to-resolve inefficiencies and therefore achieve better code performance.
+Although profiling follows the same principles of any other software project, the purpose of this document is to provide profiling samples for the most common scenarios in MLOps/Data Science projects.
+
+Below are some common scenarios in MLOps/Data Science projects, along with suggestions on how to profile them.
+
+- [Profiling Data Science and MLOps Code](#profiling-data-science-and-mlops-code)
```
[the remark this answers is not in the document]
Comment (vJenny) [reply docs/machine-learning/ml-profiling.md]: Updated, great idea, thanks!

File: docs/machine-learning/ml-profiling.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
+# Profiling Data Science and MLOps Code
+
+Data Science projects, especially the ones that involve Deep Learning techniques, usually are very resource intensive. One model training iteration might be multiple hours long. Although large data volumes processing genuinely takes time, minor bugs and nonoptimal implementation of some functional pieces might cause extra resources consumption.
```
[the remark this answers is not in the document]
Comment (vJenny) [reply docs/machine-learning/ml-profiling.md]: Done, thanks!

File: docs/machine-learning/ml-profiling.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+  - [PyTorch model training profiling](#pytorch-model-training-profiling)
+  - [Azure Machine Learning pipeline profiling](#azure-machine-learning-pipeline-profiling)
+
+## Generic Python profiling
+
+Usually an MLOps/Data Science solution contains plain Python code serving different purposes (e.g., data processing) along
+with specialized model training code. Whereas many Machine Learning frameworks provide their own profiler,
+sometimes it is also very useful to profile the whole solution.
+
+There are two types of profilers: deterministic (all events are tracked) and statistical (sampling with regular intervals). In this
```
[the remark this answers is not in the document]
Comment (vJenny) [reply docs/machine-learning/ml-profiling.md:20]: Done, thanks!

File: docs/machine-learning/ml-profiling.md [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+- [Profiling Data Science and MLOps Code](#profiling-data-science-and-mlops-code)
+  - [Generic Python profiling](#generic-python-profiling)
+  - [PyTorch model training profiling](#pytorch-model-training-profiling)
+  - [Azure Machine Learning pipeline profiling](#azure-machine-learning-pipeline-profiling)
+
+## Generic Python profiling
+
+Usually an MLOps/Data Science solution contains plain Python code serving different purposes (e.g., data processing) along
+with specialized model training code. Whereas many Machine Learning frameworks provide their own profiler,
+sometimes it is also very useful to profile the whole solution.
```
[the remark this answers is not in the document]
Comment (vJenny) [reply docs/machine-learning/ml-profiling.md]: Done, thanks!

**Review by omri374 — APPROVED** [review]
Looks great! Added a few minor comments.

File: docs/machine-learning/ml-profiling.md (hunk 7 above)
Comment (vJenny) [reply docs/machine-learning/ml-profiling.md]: Thanks! Great suggestion

**Pull request merged**